lib/collection/src/shards/local_shard/shard_ops.rs

```diff
@@ -247,15 +247,21 @@ impl ShardOperation for LocalShard {
             let update_sender = self.update_sender.load();
             let channel_permit = update_sender.reserve().await?;
 
-            let (operation_id, wal_lock) = self.wal.lock_and_write(&mut operation).await?;
+            // TODO:
+            //
+            // - Apply operation, only if it's accepted and written into the WAL...
+            // - ...or *propagate rejection to the sender node*, if it's rejected!
+
+            // It is *critical* to hold `_wal_lock` while sending operation to the update handler!
+            let (operation_id, _wal_lock) = self.wal.lock_and_write(&mut operation).await?;
 
             channel_permit.send(UpdateSignal::Operation(OperationData {
                 op_num: operation_id,
                 operation: operation.operation,
                 sender: callback_sender,
                 wait,
             }));
-            drop(wal_lock);
```

Comment:
I put it here so we don't make the same mistake again :smile:

Reply:
Ok, I kinda get your point now, but, IMO, it's not very effective and this part simply requires an explanation comment (with a lot of exclamation marks in it).

It's still super easy to just look at this code, see the `drop`, think "WTF is this here for?" and just remove the `drop` and refactor the `wal_lock` away same way we did earlier.

If we add the comment, then it will explain that the `drop` is *intentional*, and we *have to* hold the `wal_lock` for the duration of `send`... but at this point there's not much use in doing explicit `drop`, IMO.

Reply:
If I see a `drop` I always think: it must be there for a reason, lets be very careful about it.

Comments don't do the same to me while refactoring.

But anyway, fine with both.

Reply:
Hmmm...... I usually do as well, but in this case it's _kinda_ obvious that it's the end of the scope, so I usually assume it's just a leftover from a refactoring or something.